plugins/formal-verification/formver.plugin/src/org/jetbrains/kotlin/formver/reporting/VerifierErrorInterpreter.kt

```diff
@@ -24,122 +19,55 @@ import org.jetbrains.kotlin.formver.viper.errors.VerificationError
+        is SourceRole.ParamFunctionLeakageCheck -> LeakingLambdaReporter(error)
+        is SourceRole.ConditionalEffect -> ConditionalEffectReporter(sourceRole)
+        else -> DefaultErrorReporter(error)
     }
 
-    private val EventOccurrencesRange.asUserFriendlyMessage: String
-        get() = when (this) {
-            EventOccurrencesRange.AT_MOST_ONCE -> "at most once"
-            EventOccurrencesRange.EXACTLY_ONCE -> "exactly once"
-            EventOccurrencesRange.AT_LEAST_ONCE -> "at least once"
-            EventOccurrencesRange.MORE_THAN_ONCE -> "more than once"
-            else -> TODO("Unreachable")
-        }
-
-    private val SourceRole.ReturnsEffect.asUserFriendlyMessage: String
-        get() = when (this) {
-            is SourceRole.ReturnsEffect.Bool -> if (bool) "false" else "true"
-            is SourceRole.ReturnsEffect.Null -> if (negated) "null" else "non-null"
-            else -> TODO("Unreachable")
+private fun DiagnosticReporter.reportVerificationError(
+    source: KtSourceElement?,
+    error: VerificationError,
+    errorStyle: ErrorStyle,
+    context: CheckerContext,
+) {
+    when (errorStyle) {
+        ErrorStyle.USER_FRIENDLY -> {
+            val reporter = fetchVerificationErrorReporter(error)
+            reporter.report(this, source, context)
         }
-
-    private fun SourceRole.ReturnsEffect.describe(): String = when (this) {
-        is SourceRole.ReturnsEffect.Bool, is SourceRole.ReturnsEffect.Null -> "a $this value is returned"
-        SourceRole.ReturnsEffect.Wildcard -> "the function returns"
-    }
-
-    private fun SourceRole.Condition.describe(): String = when (this) {
-        is SourceRole.FirSymbolHolder -> firSymbol.rendered
-        is SourceRole.Condition.Constant -> literal.toString()
-        is SourceRole.Condition.Negation -> "!${arg.describe()}"
-        is SourceRole.Condition.Conjunction -> "(${lhs.describe()} && ${rhs.describe()})"
-        is SourceRole.Condition.Disjunction -> "(${lhs.describe()} || ${rhs.describe()})"
-        is SourceRole.Condition.IsNull -> buildString {
-            append(targetVariable.rendered)
-            when (negated) {
-                true -> append(" != ")
-                false -> append(" == ")
-            }
-            append("null")
+        ErrorStyle.ORIGINAL_VIPER -> {
+            DefaultErrorReporter(error).report(this, source, context)
         }
-        is SourceRole.Condition.IsType -> buildString {
-            append(targetVariable.rendered)
-            when (negated) {
-                true -> append(" !is ")
-                false -> append(" is ")
+        ErrorStyle.BOTH -> {
+            val reporter = fetchVerificationErrorReporter(error)
+            reporter.report(this, source, context)
+            // Avoid duplicate warnings if we do not have a specific reporter for the error.
+            if (reporter !is DefaultErrorReporter) {
+                DefaultErrorReporter(error).report(this, source, context)
```

Comment:
This is a bit clunky.  Why not the following: try to get a user friendly error message, then if that is null, fall back on the default reporter.  You don't even need to handle the logic here; you can make a helper function that takes the `ErrorStyle` and returns all relevant errors.